#159204728 Completes pagination support on all articles, articles belonging to an author and single article endpoints

chukwuemekachm opened the pull request:
#### What does this PR do?
- Completes pagination support on all articles, articles belonging to an author and single article endpoints
#### Description of Task to be completed?
- Write unit tests for the endpoints
- Create the appropriate controllers and methods
- Use POSTMAN to test the endpoints
#### How should this be manually tested?
- Clone the repository
- Start up the command line /Terminal
- Change to the repo directory
- Run `npm install`
- Run `npm run migrate`
- Run `npm run dev`
- Use POSTMAN to create a GET request to `/api/articles?limit=2&offset=2`
- Use POSTMAN to create a GET request to `/api/users/:userId/articles?limit=2&offset=2`
- Use POSTMAN to create a GET request to `/api/articles/:slug`
#### Any background context you want to provide?
- Create a Postrgresql database before you start
#### What are the relevant pivotal tracker stories?
[#159204728](https://www.pivotaltracker.com/story/show/159204728)
#### Screenshots (if appropriate)

<img width="1440" alt="screen shot 2018-08-14 at 8 13 25 am" src="https://user-images.githubusercontent.com/33798252/44077859-de783b1a-9f9c-11e8-802b-43a8fb864a60.png">
<img width="1440" alt="screen shot 2018-08-14 at 8 14 10 am" src="https://user-images.githubusercontent.com/33798252/44077860-dea3c514-9f9c-11e8-970f-6629183b5006.png">

#### Questions:
- None

Commit:
Feat(article): rating articles

- add unit tests
- add articles route for article endpoints
- add rating model and migrations
- add rating method to the Articles controller class

[Delivers #159204722]

yomigeek commented on server/controllers/ArticleController.js:
```diff
@@ -13,4 +22,106 @@ export default class ArticleController {
+      return res.status(404).json({
+        status: 'fail',
+        errors: {
+          article: [`Article with slig: ${slug} not found.`],
```
Good work @chukwuemekachm, I think you needed to type slug not 'slig'.

chukwuemekachm committed:
feat(pagination): Delivers Pagination on articles

- Write unit tests
- Modify the Tags and Article relationship
- Add queryHelper.js file
- Add seeds to article seeding file
- Add paginationValidation in ArticleValidation.js
- Add getAllArticles method in ArticleController.js
- Add getUserArticles method in ArticleController.js
- Add getSingleArticle method in ArticleController.js
- Add sendPaginationResponse in ArticleController
- Test endpoints with POSTMAN

[Delivers #159204728]

chukwuemekachm committed:
feat(pagination): Rebase feature on staging branch

- Write tests
- Add ratings in the query for all articles
- Add ratings and rater to the query for a single article

chukwuemekachm committed:
docs(pagination): resolves Hound CI violations

- resolve all Hound CI violations

Oluwafayokemi reviewed (approved):
Nice Job Chima! Well done.

Veraclins commented on server/controllers/ArticleController.js:
```diff
@@ -161,4 +167,111 @@ export default class ArticleController {
+      });
+    }
+    // Returns a 404 when a list of user articles are requested but not found
+    if (userId) {
```
Hi, @chukwuemekachm good job here. I particularly love how you simplified the whole process with helper functions. I am only consigned about this part of the code. The logic of returning an error if there is a userId is difficult to understand. I see that you are using it to differentiate the error messages for a single article and multiple articles but maybe you can consider using another method for the error situations which will use an argument that will not be misleading.

chukwuemekachm replied:
Thanks, @Veraclins, I know the logic is a bit confusing. I need the argument to assert that the request was made for a list of articles belonging to an author. If the operation fails, I then use the `userId` argument to send a personalized message to the user. But I will try to refactor that part.
Thanks.

Veraclins reviewed (changes requested):
@chukwuemekachm Good job Bro. Just look into the comment I made and you are good to go.

yomigeek reviewed (approved):
@chukwuemekachm I am satisfied with the job done, just ensure to implement the changes requested above.

chukwuemekachm committed:
feat(pagination): Implement feedback on PR

- Refactor complex logic in sendPaginationResponse method

madeofhuman commented on server/tests/singleArticle.test.js:
```diff
@@ -0,0 +1,49 @@
+        res.body.article.should.have.property('ratings');
+        res.body.article.should.have.property('tags');
+        res.body.article.User.should.be.a('object');
+        res.body.article.tags.should.be.a('array');
```
Duuuuude!!! This is a very comprehensive test! 🙌🏾
However, did you need to test all the properties? Just a couple of unique properties would have achieved the same result. Nice work, though.
